File: leetcode/3-longest-substring-without-repeating-characters.cxx
// https://leetcode.com/problems/longest-substring-without-repeating-characters

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    vector<int> memo(128, -1);
    int maxLength = 0;
    for (int start = -1, i = 0; i < s.length(); i += 1) {
      start = max(start, memo[s[i]]);
      memo[s[i]] = i;
      maxLength = max(maxLength, i - start);
    }
    return maxLength;
  }
};

File: leetcode/3.longest-substring-without-repeating-characters.rs
/*
 * @lc app=leetcode id=3 lang=rust
 *
 * [3] Longest Substring Without Repeating Characters
 */

pub struct Solution;

// @lc code=start
impl Solution {
  pub fn length_of_longest_substring(s: String) -> i32 {
    use std::collections::HashMap;
    let mut m = HashMap::new();
    let mut b = -1;
    let mut l = 0;
    for (i, x) in s.bytes().enumerate() {
      let i = i as i32;
      b = m.insert(x, i).map_or(b, |v| b.max(v));
      l = l.max(i - b);
    }
    l
  }
}
// @lc code=end
